File: android/src/main/java/com/superwallreactnative/bridges/SuperwallDelegateBridge.kt
package com.superwallreactnative.bridges

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.superwall.sdk.analytics.superwall.SuperwallEventInfo
import com.superwall.sdk.models.entitlements.SubscriptionStatus
import com.superwall.sdk.delegate.SuperwallDelegate
import com.superwall.sdk.paywall.presentation.PaywallInfo
import com.superwallreactnative.models.SuperwallEvent
import com.superwallreactnative.models.convertMapToReadableMap
import com.superwallreactnative.models.toJson
import com.superwall.sdk.models.internal.RedemptionResult
import java.net.URI
import android.net.Uri

class SuperwallDelegateBridge(
  private val reactContext: ReactContext
): SuperwallDelegate {
  override fun subscriptionStatusDidChange(
    from: SubscriptionStatus,
    to: SubscriptionStatus
  ) {
    val data = Arguments.createMap().apply {
      putString("from", from.toString())
      putString("to", to.toString())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("subscriptionStatusDidChange", data)
  }

  override fun handleSuperwallEvent(placementInfo: SuperwallEventInfo) {
    val data = Arguments.createMap().apply {
      putMap("eventInfo", placementInfo.toJson())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("handleSuperwallEvent", data)
  }

  override fun handleCustomPaywallAction(withName: String) {
    val data = Arguments.createMap().apply {
      putString("name", withName)
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("handleCustomPaywallAction", data)
  }

  override fun willDismissPaywall(withInfo: PaywallInfo) {
    val data = Arguments.createMap().apply {
      putMap("info", withInfo.toJson())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("willDismissPaywall", data)
  }

  override fun willPresentPaywall(withInfo: PaywallInfo) {
    val data = Arguments.createMap().apply {
      putMap("info", withInfo.toJson())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("willPresentPaywall", data)
  }

  override fun didDismissPaywall(withInfo: PaywallInfo) {
    val data = Arguments.createMap().apply {
      putMap("info", withInfo.toJson())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("didDismissPaywall", data)
  }

  override fun didPresentPaywall(withInfo: PaywallInfo) {
    val data = Arguments.createMap().apply {
      putMap("info", withInfo.toJson())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("didPresentPaywall", data)
  }

  override fun paywallWillOpenURL(url: URI) {
    val data = Arguments.createMap().apply {
      putString("url", url.toString())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("paywallWillOpenURL", data)
  }

  override fun paywallWillOpenDeepLink(url: Uri) {
    val data = Arguments.createMap().apply {
      putString("url", url.toString())
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("paywallWillOpenDeepLink", data)
  }

  override fun handleLog(
    level: String,
    scope: String,
    message: String?,
    info: Map<String, Any>?,
    error: Throwable?
  ) {
    val data = Arguments.createMap().apply {
      putString("level", level)
      putString("scope", scope)
      putString("message", message)
      putMap("info", info?.let { convertMapToReadableMap(it) })
      putString("error", error?.message)
    }

    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("handleLog", data)
  }

  override fun willRedeemLink(){
    try {
      reactContext
        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        .emit("willRedeemLink", null)
    } catch (e: Exception) {
      e.printStackTrace()
    }
  }

  override fun didRedeemLink(result: RedemptionResult){
    val resultJson = result.toJson()
    try {
      reactContext
        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        .emit("didRedeemLink", resultJson)
    } catch (e: Exception) {
      e.printStackTrace()
    }
  }
}

File: ios/Bridges/SuperwallDelegateBridge.swift
//
//  SuperwallDelegateBridge.swift
//  Superwall
//
//  Created by Yusuf Tör on 21/02/2024.
//

import SuperwallKit

final class SuperwallDelegateBridge: SuperwallDelegate {
  func subscriptionStatusDidChange(
    from oldValue: SubscriptionStatus,
    to newValue: SubscriptionStatus
  ) {
    let data: [String: Any] = [
      "from": oldValue.toJson(),
      "to": newValue.toJson(),
    ]
    sendEvent(withName: "subscriptionStatusDidChange", body: data)
  }

  func handleSuperwallEvent(withInfo eventInfo: SuperwallEventInfo) {
    let data: [String: Any] = ["eventInfo": eventInfo.toJson()]
    sendEvent(withName: "handleSuperwallEvent", body: data)
  }

  func handleCustomPaywallAction(withName name: String) {
    let data: [String: Any] = ["name": name]
    sendEvent(withName: "handleCustomPaywallAction", body: data)
  }

  func willDismissPaywall(withInfo paywallInfo: PaywallInfo) {
    let data: [String: Any] = ["info": paywallInfo.toJson()]
    sendEvent(withName: "willDismissPaywall", body: data)
  }

  func willPresentPaywall(withInfo paywallInfo: PaywallInfo) {
    let data: [String: Any] = ["info": paywallInfo.toJson()]
    sendEvent(withName: "willPresentPaywall", body: data)
  }

  func didDismissPaywall(withInfo paywallInfo: PaywallInfo) {
    let data: [String: Any] = ["info": paywallInfo.toJson()]
    sendEvent(withName: "didDismissPaywall", body: data)
  }

  func didPresentPaywall(withInfo paywallInfo: PaywallInfo) {
    let data: [String: Any] = ["info": paywallInfo.toJson()]
    sendEvent(withName: "didPresentPaywall", body: data)
  }

  func paywallWillOpenURL(url: URL) {
    let data: [String: Any] = ["url": url.absoluteString]
    sendEvent(withName: "paywallWillOpenURL", body: data)
  }

  func paywallWillOpenDeepLink(url: URL) {
    let data: [String: Any] = ["url": url.absoluteString]
    sendEvent(withName: "paywallWillOpenDeepLink", body: data)
  }

  func handleLog(
    level: String,
    scope: String,
    message: String?,
    info: [String: Any]?,
    error: Error?
  ) {
    let data: [String: Any] = [
      "level": level,
      "scope": scope,
      "message": message,
      "info": info,
      "error": error?.localizedDescription,
    ]
    sendEvent(withName: "handleLog", body: data)
  }

  private func sendEvent(withName name: String, body: [String: Any]) {
    SuperwallReactNative.emitter.sendEvent(withName: name, body: body)
  }

  func willRedeemLink() {
    sendEvent(withName: "willRedeemLink", body: [:])
  }

  func didRedeemLink(result: RedemptionResult) {
    sendEvent(withName: "didRedeemLink", body: result.toJson())
  }
}
